src/main/java/net/minecraftforge/fmlclient/StartupMessageRenderer.java

```diff
@@ -0,0 +1,279 @@
+            
+            if (glGetProgrami(program, GL_LINK_STATUS) != GL_TRUE) {
+                System.err.println("Program linking failed");
+                System.err.println("Program link log\n" + glGetProgramInfoLog(program));
+                throw new IllegalStateException("Program link failed");
+            }
+            
+            glDeleteShader(vertexShader);
+            glDeleteShader(fragmentShader);
+            
+            posMultiplierLocation = glGetUniformLocation(program, "posMultiplier");
+            posShiftLocation = glGetUniformLocation(program, "posShift");
+            colorInLocation = glGetUniformLocation(program, "colorIn");
+            sizeLocation = glGetUniformLocation(program, "size");
+            
+            int inputLocation = glGetAttribLocation(program, "position");
+            
+            VAO = glGenVertexArrays();
+            dataBuffer = glGenBuffers();
+            elementBuffer = glGenBuffers();
+            
+            glBindVertexArray(VAO);
+            
+            glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, elementBuffer);
+            
+            glBindBuffer(GL_ARRAY_BUFFER, dataBuffer);
+            glVertexAttribPointer(inputLocation, 2, GL_FLOAT, false, 16, 0);
+            glEnableVertexAttribArray(inputLocation);
+            
+            glBindVertexArray(0);
+            
+            glBufferData(GL_ARRAY_BUFFER, bufferLength, GL_DYNAMIC_DRAW);
+            charBuffer = MemoryUtil.memAlloc(bufferLength);
+            
+            glBindBuffer(GL_ARRAY_BUFFER, previousArrayBuffer);
+            glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, previousElementBuffer);
+            glBindVertexArray(previousVAO);
+        }
+        
+        public void close() {
+            MemoryUtil.memFree(charBuffer);
+            
+            glDeleteBuffers(elementBuffer);
+            glDeleteBuffers(dataBuffer);
+            glDeleteVertexArrays(VAO);
+            
+            glDeleteProgram(program);
+        }
+        
+        public void render(float width, float height, float textSize, boolean isDarkMode) {
+            this.height = height;
+            
+            int previousVAO = glGetInteger(GL_VERTEX_ARRAY_BINDING);
+            int previousArrayBuffer = glGetInteger(GL_ARRAY_BUFFER_BINDING);
+            int previousElementBuffer = glGetInteger(GL_ELEMENT_ARRAY_BUFFER_BINDING);
+            
+            int previousProgram = glGetInteger(GL_CURRENT_PROGRAM);
+            
+            int blendEnabled = glGetInteger(GL_BLEND);
+            int blendSrcMode = glGetInteger(GL_BLEND_SRC_ALPHA);
+            int blendDstMode = glGetInteger(GL_BLEND_DST_ALPHA);
+            
+            glEnable(GL_BLEND);
+            glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
+            
+            glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, elementBuffer);
+            glBindBuffer(GL_ARRAY_BUFFER, dataBuffer);
+            
+            glBindVertexArray(VAO);
```

Comment:
Binding a vertex array already binds the given array and element array buffers, so why are you binding them again before?

Reply:
A VAO bind does not bind buffers to the GL_ARRAY_BUFFER or GL_ELEMENT_BUFFER binding points. The only time these binding points matter is at VAO specification. The GL 4.5 Direct State Access API makes whats actually going on much clearer. The buffers are bound for when they are updated later, avoiding multiple bind calls.